File: src/integrated-scanner/integratedScanner.ts
import { ReportGenerator, type ThreatReport } from "../analyzer/reportGenerator.js";
import { RiskScorer, type RiskResult } from "../analyzer/riskScorer.js";
import { BehavioralScanner } from "../scanner/behavioralScanner.js";
import { SignatureMatcher } from "../signatures/signatureMatcher.js";
import type { ThreatSignature } from "../signatures/signatureDatabase.js";
import type { Finding, ScanTarget } from "../types.js";
import { RuleManager } from "../yara/ruleManager.js";
import type { YaraRule } from "../yara/rules.js";

/** Optional LLM classifier hook (PRD 5.4) — opt-in, async, off by default. */
export type LlmClassifier = (
  target: ScanTarget,
  priorFindings: Finding[],
) => Promise<Finding[]>;

export interface IntegratedScannerOptions {
  signatures?: ThreatSignature[];
  yaraRules?: YaraRule[];
  /** Provide to enable opt-in LLM escalation (PRD 5.4, FR). */
  llm?: LlmClassifier;
  /** Min combined score before escalating to the LLM (default 60). */
  llmEscalationThreshold?: number;
  /** Called when the opt-in LLM step throws. Defaults to `console.warn`. */
  onLlmError?: (error: unknown, target: ScanTarget) => void;
}

export interface ScanResult extends RiskResult {
  report: ThreatReport;
}

/**
 * Integrated Scanner (PRD core module `integrated-scanner`).
 *
 * Unified pipeline: Behavioral → Signature → YARA → (opt-in) LLM, fed into the
 * Risk Scoring Engine and Report Generator. This is the primary entry point a
 * mobile/host app calls.
 */
export class IntegratedScanner {
  private readonly behavioral = new BehavioralScanner();
  private readonly signatures: SignatureMatcher;
  private readonly rules: RuleManager;
  private readonly scorer = new RiskScorer();
  private readonly reporter = new ReportGenerator();
  private readonly llm?: LlmClassifier;
  private readonly llmThreshold: number;
  private readonly onLlmError: (error: unknown, target: ScanTarget) => void;

  constructor(options: IntegratedScannerOptions = {}) {
    this.signatures = new SignatureMatcher(options.signatures);
    this.rules = new RuleManager(options.yaraRules);
    this.llm = options.llm;
    this.llmThreshold = options.llmEscalationThreshold ?? 60;
    this.onLlmError =
      options.onLlmError ??
      ((err, target) =>
        console.warn(`[sync] LLM classification failed for ${target.id}:`, err));
  }

  /** Access the rule manager for custom-rule management (FR-15). */
  get ruleManager(): RuleManager {
    return this.rules;
  }

  get signatureMatcher(): SignatureMatcher {
    return this.signatures;
  }

  /** Run the full pipeline against a target and produce a scored report. */
  async scan(target: ScanTarget): Promise<ScanResult> {
    const findings: Finding[] = [
      ...this.behavioral.scan(target),
      ...this.signatures.scan(target),
      ...this.rules.scanner().scan(target),
    ];

    // Opt-in LLM escalation for ambiguous/high-stakes cases (PRD 5.4).
    if (this.llm) {
      const interim = this.scorer.score(findings);
      if (interim.score >= this.llmThreshold) {
        try {
          findings.push(...(await this.llm(target, findings)));
        } catch (err) {
          // LLM is best-effort; never let it break local detection (PRD §7),
          // but surface the failure so operators aren't blind to it.
          this.onLlmError(err, target);
        }
      }
    }

    const result = this.scorer.score(findings);
    const report = this.reporter.generate(target, result);
    return { ...result, report };
  }

  /** Synchronous local-only scan (no LLM) for tight on-device loops. */
  scanLocal(target: ScanTarget): ScanResult {
    const findings: Finding[] = [
      ...this.behavioral.scan(target),
      ...this.signatures.scan(target),
      ...this.rules.scanner().scan(target),
    ];
    const result = this.scorer.score(findings);
    const report = this.reporter.generate(target, result);
    return { ...result, report };
  }
}

File: android/engine/src/main/kotlin/com/daemonblockint/sync/engine/integrated/IntegratedScanner.kt
package com.daemonblockint.sync.engine.integrated

import com.daemonblockint.sync.engine.Finding
import com.daemonblockint.sync.engine.ScanTarget
import com.daemonblockint.sync.engine.analyzer.ReportGenerator
import com.daemonblockint.sync.engine.analyzer.RiskResult
import com.daemonblockint.sync.engine.analyzer.RiskScorer
import com.daemonblockint.sync.engine.scanner.BehavioralScanner
import com.daemonblockint.sync.engine.signatures.SignatureMatcher
import com.daemonblockint.sync.engine.signatures.DEFAULT_SIGNATURES
import com.daemonblockint.sync.engine.signatures.ThreatSignature
import com.daemonblockint.sync.engine.yara.RuleManager
import com.daemonblockint.sync.engine.yara.SOLANA_YARA_RULES
import com.daemonblockint.sync.engine.yara.YaraRule

/** Optional LLM classifier hook (PRD 5.4) — opt-in, async, off by default. */
fun interface LlmClassifier {
    suspend fun classify(target: ScanTarget, priorFindings: List<Finding>): List<Finding>
}

data class IntegratedScannerOptions(
    val signatures: List<ThreatSignature> = DEFAULT_SIGNATURES,
    val yaraRules: List<YaraRule> = SOLANA_YARA_RULES,
    val llm: LlmClassifier? = null,
    val llmEscalationThreshold: Int = 60,
    val onLlmError: (error: Throwable, target: ScanTarget) -> Unit = { err, target ->
        println("[sync] LLM classification failed for ${target.id}: $err")
    },
)

/** ScanResult extends RiskResult with the generated report. */
data class ScanResult(
    val score: Int,
    val severity: com.daemonblockint.sync.engine.Severity,
    val breakdown: RiskResult.Breakdown,
    val escalated: Boolean,
    val findings: List<Finding>,
    val report: com.daemonblockint.sync.engine.analyzer.ThreatReport,
)

/**
 * Integrated Scanner (PRD core module `integrated-scanner`).
 *
 * Unified pipeline: Behavioral → Signature → YARA → (opt-in) LLM, fed into the
 * Risk Scoring Engine and Report Generator. This is the primary entry point a
 * mobile/host app calls.
 */
class IntegratedScanner(
    private val options: IntegratedScannerOptions = IntegratedScannerOptions(),
) {
    private val behavioral = BehavioralScanner()
    private val signatures: SignatureMatcher = SignatureMatcher(options.signatures.toMutableList())
    private val rules: RuleManager = RuleManager(options.yaraRules)
    private val scorer = RiskScorer()
    private val reporter = ReportGenerator()

    /** Access the rule manager for custom-rule management (FR-15). */
    val ruleManager: RuleManager get() = rules
    val signatureMatcher: SignatureMatcher get() = signatures

    /** Run the full pipeline against a target and produce a scored report. */
    suspend fun scan(target: ScanTarget): ScanResult {
        val findings = mutableListOf<Finding>()
        findings.addAll(behavioral.scan(target))
        findings.addAll(signatures.scan(target))
        findings.addAll(rules.scanner().scan(target))

        // Opt-in LLM escalation for ambiguous/high-stakes cases (PRD 5.4).
        if (options.llm != null) {
            val interim = scorer.score(findings)
            if (interim.score >= options.llmEscalationThreshold) {
                try {
                    findings.addAll(options.llm.classify(target, findings.toList()))
                } catch (err: Throwable) {
                    options.onLlmError(err, target)
                }
            }
        }

        return finalize(target, findings)
    }

    /** Synchronous local-only scan (no LLM) for tight on-device loops. */
    fun scanLocal(target: ScanTarget): ScanResult {
        val findings = mutableListOf<Finding>()
        findings.addAll(behavioral.scan(target))
        findings.addAll(signatures.scan(target))
        findings.addAll(rules.scanner().scan(target))
        return finalize(target, findings)
    }

    private fun finalize(target: ScanTarget, findings: List<Finding>): ScanResult {
        val result = scorer.score(findings)
        val report = reporter.generate(target, result)
        return ScanResult(
            score = result.score,
            severity = result.severity,
            breakdown = result.breakdown,
            escalated = result.escalated,
            findings = result.findings,
            report = report,
        )
    }
}
